ext/native/gfx_es2/draw_text_android.cpp

```diff
@@ -0,0 +1,233 @@
+	fontHash_ = fontHash;
+	return fontHash;
+}
+
+void TextDrawerAndroid::SetFont(uint32_t fontHandle) {
+	uint32_t fontHash = fontHandle;
+	auto iter = fontMap_.find(fontHash);
+	if (iter != fontMap_.end()) {
+		curSize_ = iter->second.size / dpiScale_;
+	}
+}
+
+std::string TextDrawerAndroid::NormalizeString(std::string str) {
+	return ReplaceAll(str, "&&", "&");
+}
+
+void TextDrawerAndroid::RecreateFonts() {
+
+}
+
+void TextDrawerAndroid::MeasureString(const char *str, size_t len, float *w, float *h) {
+	std::string text(NormalizeString(std::string(str, len)));
+	jstring jstr = env_->NewStringUTF(text.c_str());
+	uint32_t size = env_->CallStaticIntMethod(cls_textRenderer, method_measureText, jstr, curSize_);
+	env_->DeleteLocalRef(jstr);
+	*w = (size >> 16) * fontScaleX_ * dpiScale_;
+	*h = (size & 0xFFFF) * fontScaleY_ * dpiScale_;
+}
+
+void TextDrawerAndroid::MeasureStringRect(const char *str, size_t len, const Bounds &bounds, float *w, float *h, int align) {
+	std::string toMeasure(NormalizeString(std::string(str, len)));
+	if (align & FLAG_WRAP_TEXT) {
+		bool rotated = (align & (ROTATE_90DEG_LEFT | ROTATE_90DEG_RIGHT)) != 0;
+		WrapString(toMeasure, toMeasure.c_str(), rotated ? bounds.h : bounds.w);
+	}
+
+	jstring jstr = env_->NewStringUTF(toMeasure.c_str());
+	uint32_t size = env_->CallStaticIntMethod(cls_textRenderer, method_measureText, jstr, curSize_);
+	env_->DeleteLocalRef(jstr);
+	*w = (size >> 16) * fontScaleX_ * dpiScale_;
+	*h = (size & 0xFFFF) * fontScaleY_ * dpiScale_;
+}
+
+void TextDrawerAndroid::DrawString(DrawBuffer &target, const char *str, float x, float y, uint32_t color, int align) {
+	using namespace Draw;
+	std::string text(NormalizeString(std::string(str)));
+	if (text.empty())
+		return;
+	JNIEnv *env;
+	int result = javaVM->GetEnv((void **)&env, JNI_VERSION_1_6);
+	assert(env == env_);
+
+	uint32_t stringHash = hash::Fletcher((const uint8_t *)text.data(), text.size());
+	uint32_t entryHash = stringHash ^ fontHash_ ^ (align << 24);
+
+	target.Flush(true);
+
+	TextStringEntry *entry;
+
+	auto iter = cache_.find(entryHash);
+	if (iter != cache_.end()) {
+		entry = iter->second.get();
+		entry->lastUsedFrame = frameCount_;
+		draw_->BindTexture(0, entry->texture);
+	} else {
+		jstring jstr = env_->NewStringUTF(text.c_str());
+		int len = (int)env_->GetStringUTFLength(jstr);
+		ILOG("UTF len: %d", len);
+		uint32_t size = env_->CallStaticIntMethod(cls_textRenderer, method_measureText, jstr, curSize_);
+		int imageWidth = (size >> 16);
+		int imageHeight = (size & 0xFFFF);
+		jintArray imageData = (jintArray)env_->CallStaticObjectMethod(cls_textRenderer, method_renderText, jstr, curSize_);
+		env_->DeleteLocalRef(jstr);
+
+		entry = new TextStringEntry();
+		entry->bmWidth = entry->width = imageWidth;
+		entry->bmHeight = entry->height = imageHeight;
+		entry->lastUsedFrame = frameCount_;
+
+		TextureDesc desc{};
+		desc.type = TextureType::LINEAR2D;
+		desc.format = Draw::DataFormat::B4G4R4A4_UNORM_PACK16;
```

Comment:
We don't need to worry about this ever being unsupported on Android, right?

-[Unknown]